blog-posts/cloudfront-cache-efficiency.md

```diff
@@ -0,0 +1,154 @@
+4. 60 seconds after the first requests, CloudFront clears the cache. The request hits the lambda
+   function, so it takes around 700ms. After that requests, the same procedure starts again.
+
+What is wrong here?
+
+1. The browser should cache every request for 10 seconds. But it only caches the first requests for the `max-age` and #
+doesn't cache anything for the rest of the remaining `s-maxage` time.
+2. After 10 seconds the browser requests the same resource again. As there was no change in the response body, 
+CloudFront should only return `304`. Instead, the whole body gets return with a status `200`.
+
+Those 2 problems will result in
+
+- Additional **traffic you have to pay** between the browser and your CDN.
+- A **slower website**, as the browser has to wait for the request to complete. This problem increases if your response
+body gets larger or the network speed of your customer decreases (a.e. on mobile plans).
+- Additional **traffic your customer has to pay** if the data is not free (a.e. on mobile plans).
+
+## Fixing the browsers cache behavior
+
+We spend hours and hours figuring out why the browser only caches the response for the `max-age` time, and then ignores 
+the cache for the rest of the `s-maxage` time. As it turns out, we were not the first ones how stumbled upon this problem:
+
+- [https://www.cdnplanet.com/blog/cloudfront-cachability-date-header/](https://www.cdnplanet.com/blog/cloudfront-cachability-date-header/)
+- [https://forums.aws.amazon.com/thread.jspa?messageID=807813](https://forums.aws.amazon.com/thread.jspa?messageID=807813)
+- [https://stackoverflow.com/a/61493383/3141881](https://stackoverflow.com/a/61493383/3141881)
+
+It turns out that the root of the problem is the `date` header returned by CloudFront. It doesn’t change on upcoming
+requests. As caching times in the `Cache-Control` header are relative to the response `date` header, only the first 
+request by the browser get's cached for the `max-age` time. 
+
+> CDNs usually don't cache the `date` header, it's only CloudFront that does it.
+
+There is no easy way to fix this within some CloudFront settings. Only recently it is possible to solve this problem at
+all: With a custom CloudFront Function. CloudFront Functions are small Javascript functions that are executed with every
+request. You can use them to do simple changes to the request or response object. Read more about the possibilities and
+limit in the official AWS docs
+here: [https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-javascript-runtime-features.html](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-javascript-runtime-features.html)
+
+We can use those functions to update the `date` header with every request.
+
+```js
+function handler(event) {
+  var response = event.response;
+
+  response.headers['date'] = {
+    value: new Date().toUTCString()
+  };
+  delete response.headers['age'];
+
+  return response;
+}
+```
```

Comment:
I immediately wondered: Does this have any impact on billing or performance if this is run for every resources?